prov/usnic/src/usdf_eq.c

```diff
@@ -327,60 +276,39 @@ usdf_eq_sread_fd(struct fid_eq *feq, uint32_t *event, void *buf, size_t len,
 
 	eq = eq_ftou(feq);
 
-	/* block awaiting event */
+	/* Setup poll context to block until the FD becomes readable. */
 	pfd.fd = eq->eq_fd;
 	pfd.events = POLLIN;
+
 retry:
 	ret = poll(&pfd, 1, timeout);
-	if (ret < 0) {
+	if (ret < 0)
 		return -errno;
-	} else if (ret == 0) {
+	else if (ret == 0)
 		return -FI_EAGAIN;
-	}
-
-	pthread_spin_lock(&eq->eq_lock);
 
-	/* did someone steal our event? */
-	if (usdf_eq_empty(eq)) {
-		pthread_spin_unlock(&eq->eq_lock);
+	ret = _usdf_eq_read(eq, event, buf, len, flags);
+	if (ret == -FI_EAGAIN)
 		goto retry;
-	}
-
-	if (usdf_eq_error(eq)) {
-		ret = -FI_EAVAIL;
-		goto done;
-	}
-
-	/* consume the event in eventfd */
-	ret = read(eq->eq_fd, &val, sizeof(val));
-	if (ret != sizeof(val)) {
-		ret = -FI_EIO;
-		goto done;
-	}
-
-	ret = usdf_eq_read_event(eq, event, buf, len, flags);
 
-done:
-	pthread_spin_unlock(&eq->eq_lock);
 	return ret;
 }
 
-static ssize_t
-usdf_eq_write_fd(struct fid_eq *feq, uint32_t event, const void *buf,
-		size_t len, uint64_t flags)
+ssize_t usdf_eq_write_internal(struct usdf_eq *eq, uint32_t event,
+		const void *buf, size_t len, uint64_t flags)
 {
-	struct usdf_eq *eq;
-	uint64_t val;
+	uint64_t val = 1;
 	int ret;
 	int n;
 
-	USDF_DBG_SYS(EQ, "\n");
-
-	eq = eq_ftou(feq);
+	USDF_DBG_SYS(EQ, "event=%#" PRIx32 " flags=%#" PRIx64 "\n", event,
```

Comment:
The '#' in the format string, when used with a '0x' modifier, will print a leading '0x' only if the value is non-zero. If it's zero then it will just print 0.

Example:

```
#include <stdio.h>
#include <inttypes.h>

int main(void) {
        uint32_t a = 0x400;
        uint32_t b = 0;

        printf("Hello %#" PRIx32 " is zero while %#" PRIx32 " is not.\n",
                        b, a);
        printf("Hello %" PRIx32 " is zero while %" PRIx32 " is not.\n",
                        b, a);

        return 0;
}
```

prints this:

```
Hello 0 is zero while 0x400 is not.
Hello 0 is zero while 400 is not.
```